_posts/2019-09-03-faster-strings-cpp-parquet.md

```diff
@@ -0,0 +1,238 @@
+<img src="{{ site.base-url }}/img/20190903-parquet-dictionary-column-chunk.png"
+     alt="Internal ColumnChunk structure"
+     width="80%" class="img-responsive">
+</div>
+
+# Faster reading and writing of dictionary-encoded data
+
+When reading a Parquet file, the dictionary-encoded portions are usually
+materialized to their non-dictionary-encoded form, causing binary or string
+values to be duplicated in memory. So an obvious (but not trivial) optimization
+is to skip this "dense" materialization. There are several issues to deal with:
+
+* A Parquet file often contains multiple ColumnChunks for each semantic column,
+  and the dictionary values may be different in each ColumnChunk
+* We must gracefully handle the "fall back" portion which is not
+  dictionary-encoded
+
+We pursued several avenues to help with this:
+
+* Allowing each `arrow::DictionaryArray` to have a different dictionary
+  (before, the dictionary was part of the `DictionaryType`, which caused
+  problems)
+* We enabled the Parquet dictionary indices to be directly written into an
+  Arrow `DictionaryBuilder` without rehashing the data
+* When decoding a ColumnChunk, we first append the dictionary values and
+  indices into an Arrow `DictionaryBuilder`, and when we encounter the "fall
+  back" portion we use a hash table to convert those values to
+  dictionary-encoded form
+* We override the "fall back" logic when writing a ColumnChunk from an
+  `arrow::DictionaryArray` so that reading such data back is more efficient
+
+All of these things together have produced some excellent performance results
+that we will detail below.
+
+The other class of optimizations we implemented was removing an abstraction
+layer between the low-level Parquet column data encoder and decoder classes and
+the Arrow columnar data structures. This involves:
+
+* Adding `ColumnWriter::WriteArrow` and `Encoder::Put` methods that accept
+  `arrow::Array` objects directly
+* Adding `ByteArrayDecoder::DecodeArrow` method to decode binary data directly
+  into an `arrow::BinaryBuilder`.
+
+While the performance improvements from this work are less dramatic than for
+dictionary-encoded data, they are still meaningful in real-world applications.
+
+# Performance Benchmarks
+
+We ran some benchmarks comparing Arrow 0.12.1 with the current master
+branch. We construct two kinds of Arrow tables with 10 columns each:
+
+* "Low cardinality" and "high cardinality" variants. The "low cardinality" case
+  has 1,000 unique string values of 32-bytes each. The "high cardinality" has
+  100,000 unique values
+* "Dense" (non-dictionary) and "Dictionary" variants
+
+[See the full benchmark script.][11]
+
+We show both single-threaded and multithreaded read performance. The test
+machine is an Intel i9-9960X using gcc 8.3.0 (on Ubuntu 18.04) with 16 physical
+cores and 32 virtual cores.
+
+First, the writing benchmarks:
+
+<div align="center">
+<img src="{{ site.base-url }}/img/20190903_parquet_write_perf.png"
```

Comment:
Could you add units to the x-axis?  I'm assuming the time reported is seconds?  Or if it is unitless could to mention what the time is normalized to?

Reply:
Added a sentence to say it's all seconds. I'm not sure the absolute number is that meaningful here